### PR TITLE
Enable authorized access to POST to actuator endpoints

### LINKED ISSUE #210: Allow authorized user to post to `/actuator/bindings`
Related to https://github.com/spring-cloud/spring-cloud-dataflow/issues/4758  and (private) https://github.com/pivotal/scdf-pro-ui/issues/69  We need the ability to grant a user with admin privileges to post to exposed actuator endpoints, notably `bindings`.

### PR DESCRIPTION
**Pull request by dturanski**
Resolves #210

### CODE REVIEW

File: applications/stream-applications-core/stream-applications-security-common/src/main/java/org/springframework/cloud/stream/app/security/common/AppStarterWebFluxSecurityAutoConfiguration.java [hunk 1]
```diff
@@ -49,19 +61,61 @@
...
 		if (!securityProperties.isEnabled()) {
 			http.authorizeExchange()
 					.anyExchange()
 					.permitAll();
 		}
 		else {
-			http.authorizeExchange().pathMatchers("/actuator", "/actuator/health", "/actuator/info", "/actuator/bindings")
+			http.authorizeExchange()
+					.pathMatchers(HttpMethod.POST, postToActuatorPath).hasRole("ADMIN")
+					.pathMatchers("/actuator", "/actuator/health", "/actuator/info", "/actuator/bindings")
```
Comment (artembilan) [inline applications/stream-applications-core/stream-applications-security-common/src/main/java/org/springframework/cloud/stream/app/security/common/AppStarterWebFluxSecurityAutoConfiguration.java:118]: Don't we need to use the same `EndpointRequest.to("health", "info", "bindings")` over here, too?
I guess that `webEndpointProperties.getBasePath()` could be used otherwise, but I believe to keep it as an `/actuator` prefix is wrong.

**Review by artembilan — CHANGES_REQUESTED** [review]

File: applications/stream-applications-core/stream-applications-security-common/src/main/java/org/springframework/cloud/stream/app/security/common/AppStarterWebSecurityAutoConfiguration.java [hunk 2]
```diff
@@ -47,29 +59,71 @@
...
+
 	@Bean
 	WebSecurityConfigurerAdapter appStarterWebSecurityConfigurerAdapter(
 			AppStarterWebSecurityAutoConfigurationProperties securityProperties) {
 
 		return new WebSecurityConfigurerAdapter() {
-
 			@Override
 			protected void configure(HttpSecurity http) throws Exception {
+				AntPathRequestMatcher postToActuator = new AntPathRequestMatcher(
```
Comment (onobc) [inline applications/stream-applications-core/stream-applications-security-common/src/main/java/org/springframework/cloud/stream/app/security/common/AppStarterWebSecurityAutoConfiguration.java:102]: One caveat of building up the path like this is that it does not handle the case where the servlet container is configured w/ a context path other than `/` (eg. '/myapp`). The matcher that EndpointRequest [ultimately leverages](https://github.com/spring-projects/spring-boot/blob/4494d3a805fa686b7dc6c56f22d5a2dd1ad978c8/spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/security/servlet/SecurityRequestMatchersManagementContextConfiguration.java#L53) does respect the path. 

If we care about that we can do something like I did in my prototype [here](https://github.com/onobc/playground/blob/main/actuator-security/src/main/java/com/example/actuator/security/WebSecurityConfig.java#L48). However, me and @dturanski discussed this and it is probably an edge case that we would not need to worry about.

File: applications/stream-applications-core/stream-applications-security-common/src/test/java/org/springframework/cloud/stream/app/security/common/AbstractSecurityCommonTests.java [hunk 3]
```diff
@@ -33,8 +39,16 @@
 	@Autowired
 	protected TestRestTemplate restTemplate;
 
+
+
 	@SpringBootApplication
+	@Import(TestChannelBinderConfiguration.class)
```
Comment (onobc) [inline applications/stream-applications-core/stream-applications-security-common/src/test/java/org/springframework/cloud/stream/app/security/common/AbstractSecurityCommonTests.java:45]: Nice add on the test!

File: applications/stream-applications-core/stream-applications-security-common/src/main/java/org/springframework/cloud/stream/app/security/common/AppStarterWebFluxSecurityAutoConfiguration.java (hunk 1 above)
Comment (onobc) [reply applications/stream-applications-core/stream-applications-security-common/src/main/java/org/springframework/cloud/stream/app/security/common/AppStarterWebFluxSecurityAutoConfiguration.java:118]: This is a good point. The imperative version does not have the "actuator" paths. It would be a simple change to bring this inline w/ the mvc version. The api does require a string path and we can inject in PathMappedEnpdoints bean do something like this: 
```java
pathMappedEndpoints.getPath(EndpointId.of("health"))
```
to get each path.

File: applications/stream-applications-core/stream-applications-security-common/src/main/java/org/springframework/cloud/stream/app/security/common/AppStarterWebSecurityAutoConfiguration.java [hunk 4]
```diff
@@ -47,29 +59,71 @@
...
+	AuthenticationProvider adminAuthenticationProvider(
+			AppStarterWebSecurityAutoConfigurationProperties securityProperties) {
+		return new AuthenticationProvider() {
+			@Override
+			public Authentication authenticate(Authentication authentication) throws AuthenticationException {
+				if (securityProperties.isEnabled() &&
+						StringUtils.hasText(securityProperties.getAdminUser()) &&
+						securityProperties.getAdminUser().equals(authentication.getName()) &&
+						StringUtils.hasText(securityProperties.getAdminPassword()) &&
+						securityProperties.getAdminPassword().equals(authentication.getCredentials().toString())) {
```
Comment (onobc) [inline applications/stream-applications-core/stream-applications-security-common/src/main/java/org/springframework/cloud/stream/app/security/common/AppStarterWebSecurityAutoConfiguration.java:79]: Will `credentials` always be non-null?

File: applications/stream-applications-core/stream-applications-security-common/src/test/java/org/springframework/cloud/stream/app/security/common/ReactiveSecurityEnabledAuthorizedAccessTests.java [hunk 5]
```diff
@@ -40,14 +42,19 @@
 @TestPropertySource(properties = {
 		"spring.main.web-application-type=reactive",
 		"management.endpoints.web.exposure.include=*",
+		"management.info.env.enabled=true",
```
Comment (onobc) [inline applications/stream-applications-core/stream-applications-security-common/src/test/java/org/springframework/cloud/stream/app/security/common/ReactiveSecurityEnabledAuthorizedAccessTests.java:45]: Ah yeh, disabled since 2.6.3

File: applications/stream-applications-core/stream-applications-security-common/src/test/java/org/springframework/cloud/stream/app/security/common/ReactiveSecurityDisabledManagementSecurityEnabledTests.java [hunk 6]
```diff
@@ -38,8 +37,8 @@
 		"spring.main.web-application-type=reactive",
 		"spring.cloud.streamapp.security.enabled=false",
 		"management.endpoints.web.exposure.include=health,info,bindings,env",
+		"management.info.env.enabled=true",
 		"info.name=MY TEST APP"})
```
Comment (onobc) [inline applications/stream-applications-core/stream-applications-security-common/src/test/java/org/springframework/cloud/stream/app/security/common/ReactiveSecurityDisabledManagementSecurityEnabledTests.java:41]: For posterity, should we add a test that shows we can invoke the POST w/o creds when security is disabled?

Comment (dturanski) [reply applications/stream-applications-core/stream-applications-security-common/src/test/java/org/springframework/cloud/stream/app/security/common/ReactiveSecurityDisabledManagementSecurityEnabledTests.java:41]: Good catch. I had the same thought.

File: applications/stream-applications-core/stream-applications-security-common/src/main/java/org/springframework/cloud/stream/app/security/common/AppStarterWebSecurityAutoConfiguration.java (hunk 4 above)
Comment (dturanski) [reply applications/stream-applications-core/stream-applications-security-common/src/main/java/org/springframework/cloud/stream/app/security/common/AppStarterWebSecurityAutoConfiguration.java:79]: I believe that's the case once it gets here, but I'll add a check.

File: applications/stream-applications-core/stream-applications-security-common/src/main/java/org/springframework/cloud/stream/app/security/common/AppStarterWebSecurityAutoConfiguration.java [hunk 7]
```diff
@@ -126,4 +127,38 @@ protected void configure(HttpSecurity http) throws Exception {
...
+		 * For example: <pre class="code">
+		 * EndpointRequest.to(HttpMethod.POST, "loggers")
+		 * </pre>
+		 * @param httpMethod the http method to include
+		 * @param endpoints the endpoints to include
+		 * @return the configured {@link RequestMatcher}
+		 */
+		static RequestMatcher to(HttpMethod httpMethod, String... endpoints) {
+			final EndpointRequest.EndpointRequestMatcher matcher = EndpointRequest.to(endpoints);
+			return (request) -> {
```
Comment (onobc) [inline applications/stream-applications-core/stream-applications-security-common/src/main/java/org/springframework/cloud/stream/app/security/common/AppStarterWebSecurityAutoConfiguration.java:146]: **[TinyNit]** we could make this a bit more DRY and create a private helper that takes in a request matcher guarded by httpmethod which would then be called by both of these. Not a big deal either way though.

File: applications/stream-applications-core/stream-applications-security-common/src/test/java/org/springframework/cloud/stream/app/security/common/SecurityEnabledAuthorizedAccessTests.java [hunk 8]
```diff
@@ -73,6 +73,7 @@ public void testHealthEndpoint() {
 	@Test
 	@SuppressWarnings("rawtypes")
 	public void testInfoEndpoint() {
+		restTemplate.getRestTemplate().getInterceptors().clear();
```
Comment (onobc) [inline applications/stream-applications-core/stream-applications-security-common/src/test/java/org/springframework/cloud/stream/app/security/common/SecurityEnabledAuthorizedAccessTests.java:76]: So this test is really doing a "when security is enabled and I am logged in I can access X,Y,Z" so I think we should not clear out the interceptors.

Comment (onobc) [reply applications/stream-applications-core/stream-applications-security-common/src/test/java/org/springframework/cloud/stream/app/security/common/SecurityEnabledAuthorizedAccessTests.java:76]: This really brings up a good point (which is outside of the scope of this PR) that a 1-liner high level description  on each of these tests would go a long way as they are all very similiar looking and take a minute to get my head around which one is doing what.

**Review by onobc — CHANGES_REQUESTED** [review]
I think we should leave the interceptors in the test https://github.com/spring-cloud/stream-applications/pull/212/files/bbfcb601d86901fd0f58519d758d0e40f8dcaff6..b84bd0df47901c551027d72ddc2ad14e634bf4d4#r792181914

**Review by sobychacko — APPROVED** [review]

**Comment by dturanski** [comment]
merged.
